core/src/main/java/org/jruby/ir/IRBuilder.java

```diff
@@ -3980,75 +3982,77 @@ public Operand buildZArray(Variable result) {
+            scope.setUsesZSuper();
 
-                addInstr(new LabelInstr(next));
+            // Two conditions will inject an error:
+            // 1. We cannot find any method scope above the closure (e.g. module A; define_method(:a) { super }; end)
+            // 2. One of the method calls the closure is passed to is named define_method.
+            //
+            // Note: We are introducing an issue but it is so obscure we are ok with it.
+            // A method named define_method containing zsuper in a method scope which is not actually
+            // a define_method will get raised as invalid even though it should zsuper to the method.
+            addRaiseError("RuntimeError",
+                    "implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.");
+        }
 
-                // If we hit a method, this is known to always succeed
-                if (superScope instanceof IRMethod) {
-                    Operand[] args = adjustVariableDepth(getCallArgs(superScope, superBuilder), depthFromSuper);
-                    addInstr(new ZSuperInstr(scope, zsuperResult, buildSelf(), args, block, scope.maybeUsingRefinements()));
-                } //else {
-                // FIXME: Do or don't ... there is no try
-                    /* Control should never get here in the runtime */
-                    /* Should we add an exception throw here just in case? */
-                //}
+        return zsuperResult;
+    }
 
-                addInstr(new LabelInstr(allDoneLabel));
-                return zsuperResult;
-            }
-        };
+    private boolean isMethodDefine() {
+        if (methodName != null) {
+            String name = methodName.asJavaString();
 
-        return receiveBreakException(block, zsuperBuilder);
+            return "define_method".equals(name) || "define_singleton_method".equals(name);
+        }
+
+        return false;
     }
 
-    public Operand buildZSuper(ZSuperNode zsuperNode) {
-        if (scope.isModuleBody()) return buildSuperInScriptBody();
+    private void addRaiseError(String id, String message) {
+        Ruby runtime = scope.getManager().getRuntime();
+        Operand exceptionClass = searchModuleForConst(new ObjectClass(), runtime.newSymbol(id));
+        Operand exception = addResultInstr(CallInstr.create(scope,
+                createTemporaryVariable(),
+                runtime.newSymbol("new"),
+                exceptionClass,
+                new Operand[]{new StringLiteral(message)},
+                null));
+        Operand kernel = searchModuleForConst(new ObjectClass(), runtime.newSymbol("Kernel"));
+        addResultInstr(CallInstr.create(scope,
+                createTemporaryVariable(),
+                runtime.newSymbol("raise"),
+                kernel,
+                new Operand[] { exception },
+                null));
```

Comment:
You could just call this as `Kernel.raise WhateverException, "message"` rather than doing the call to `new`.

Reply:
sure.  I can make this change...less generated instrs + bytecode.